keps/104-StartupPolicy/README.md

```diff
@@ -0,0 +1,407 @@
+            containers:
+            - name: messagequeue
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 100
+  - name: driver
+    replicas: 2
+    template:
+      spec:
+        backoffLimit: 0 
+        completions: 2
+        parallelism: 2
+        template:
+          spec:
+            containers:
+            - name: driver
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10
+  - name: worker
+    replicas: 2
+    template:
+      spec:
+        backoffLimit: 0 
+        completions: 2
+        parallelism: 2
+        template:
+          spec:
+            containers:
+            - name: worker
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10
+```
+
+### Risks and Mitigations
+
+To avoid the JobSet become a half-baked workflow manager, we will not allow for starting particular replicated jobs.
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### API Proposal
+
+```golang
+const (
+ ReplicatedJobStart string = "jobset.sigs.k8s.io/replicated-job-to-start"
+)
+```
+
+```golang
+type StartupPolicyOrderOptions string
+
+const (
+ // This is the default setting
+ // AnyOrder means that we will start jobs without any specific order.
+ AnyOrder StartupPolicyOrderOptions = "AnyOrder"
+ // InOrder starts the jobs in order that they are listed.
+ InOrder StartupPolicyOrderOptions = "InOrder"
+)
+
+type StartupPolicy struct {
+ // StartupPolicyOrder is an enum of different options for StartupPolicy
+ // AnyOrder means to start replicated jobs in any order.
+ // InOrder means to start them as they are listed in the JobSet.
+ // +kubebuilder:validation:Enum=AnyOrder;InOrder
+ StartupPolicyOrder StartupPolicyOrderOptions `json:"startupPolicyOrder"`
+}
+```
+
+### Implementation
+
+We have three cases where startupPolicy applies:
+
+- when the jobset is first created
+- when the jobset is restarted after a failure (with restart or recreate all failure policy)
+- when the jobset is resumed after suspension
+
+In cases 1 and 2, the operator controls the order of creating the jobs, and so it creates the replicatedJob only when its turn comes in the startup sequence.
+In case 3, the operator controls the order of unsuspending the jobs, and so it unsuspends the replicatedJob only when its turn comes in the startup sequence.
+
+We use the annotation `jobset.sigs.k8s.io/replicated-job-to-start` to track which replicatedJob is next to create or unsuspend.
```

Comment:
Why do you need this?
In which object is it stored?
How/when is it updated?

Reply:
https://github.com/kubernetes-sigs/jobset/pull/244#discussion_r1424422189

Reply:
The answers to these questions should be in the KEP.

Reply:
Thanks for pointing that out. I added a little mention of this.